src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+        const value = entry.value_ptr.*;
+        try testing.expectEqual(key, value);
+    }
+}
+
+test "hashmap can map f64 to segment metadata array list" {
+    const allocator = testing.allocator;
+    var f64_metadata_hash_map = HashMap(
+        f64,
+        ArrayList(SegmentMetadata),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ).init(allocator);
+    defer {
+        var iterator = f64_metadata_hash_map.iterator();
+        while (iterator.next()) |entry| {
+            entry.value_ptr.*.deinit();
+        }
+        f64_metadata_hash_map.deinit();
+    }
+
+    var f64_usize_hash_map = HashMap(
+        f64,
+        usize,
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ).init(allocator);
+    defer f64_usize_hash_map.deinit();
+
+    var rnd = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
+
+    for (0..200) |_| {
+        const rand_number = @floor((rnd.random().float(f64) - 0.5) * 100) / 10;
+
+        const count_map_result = try f64_usize_hash_map.getOrPut(rand_number);
+        if (!count_map_result.found_existing) {
+            count_map_result.value_ptr.* = 0;
+        }
+        count_map_result.value_ptr.* += 1;
+
+        const metadata_map_result = try f64_metadata_hash_map.getOrPut(rand_number);
+        if (!metadata_map_result.found_existing) {
+            metadata_map_result.value_ptr.* = ArrayList(SegmentMetadata).init(allocator);
+        }
+        try metadata_map_result.value_ptr.*.append(SegmentMetadata{
+            .start_time = count_map_result.value_ptr.*,
+            .interception = rand_number,
+            .lower_bound_slope = rand_number,
+            .upper_bound_slope = rand_number,
+        });
+    }
+
+    var iterator_map = f64_metadata_hash_map.iterator();
+    while (iterator_map.next()) |entry| {
+        const expected_array_size: usize = f64_usize_hash_map.get(entry.key_ptr.*).?;
+        try testing.expectEqual(expected_array_size, entry.value_ptr.*.items.len);
+
+        for (entry.value_ptr.*.items, 1..) |item, i| {
+            try testing.expectEqual(i, item.start_time);
+            try testing.expectEqual(entry.key_ptr.*, item.lower_bound_slope);
+        }
+    }
+}
+
+test "sim-piece can compress, decompress and merge many segments with non-zero error bound" {
+    const allocator = testing.allocator;
+
+    var list_values = ArrayList(f64).init(allocator);
+    defer list_values.deinit();
+    var compressed_values = ArrayList(u8).init(allocator);
+    defer compressed_values.deinit();
+    var decompressed_values = ArrayList(f64).init(allocator);
+    defer decompressed_values.deinit();
+    var rnd = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
+
+    const error_bound: f32 = @floor(rnd.random().float(f32) * 100) / 10 + 0.1;
+
+    for (0..200) |_| {
+        // Generate floating points between numbers between 0 and 10. This will render many
```

Comment:
Unclear what render many merged segments mean.

Reply:
Fixed.

Reply:
Are the values guaranteed to create many merged segments? If so maybe replace "must" with "will".

Reply:
Fixed.